packages/core/src/components/forms/numericInput.tsx

```diff
@@ -0,0 +1,407 @@
+            const nextValue = (currValue.length > 0) ? this.getAdjustedValue(currValue, /* delta */ 0, min, max) : "";
+            this.setState({ value: nextValue });
+        }
+    }
+
+    private handleInputUpdate = (e: React.KeyboardEvent<HTMLInputElement>) => {
+        const nextValue = (e.target as HTMLInputElement).value;
+        this.setState({ shouldSelectAfterUpdate : false, value: nextValue });
+    }
+
+    private updateValue(direction: number, e: React.MouseEvent<HTMLElement> | React.KeyboardEvent<HTMLElement>) {
+        const { min, max } = this.props;
+
+        const delta = this.getDelta(direction, e);
+
+        // pretend we're incrementing from 0 if currValue isn't defined
+        const currValue = this.state.value || NumericInput.VALUE_ZERO;
+        const nextValue = this.getAdjustedValue(currValue, delta, min, max);
+
+        this.setState({ shouldSelectAfterUpdate : true, value: nextValue });
+    }
+
+    private getAdjustedValue(value: string, delta: number, min?: number, max?: number) {
+        if (!this.isValueNumeric(value)) {
+            return NumericInput.VALUE_EMPTY;
+        }
+
+        // truncate floating-point result to avoid precision issues when adding
+        // non-integer, binary-unfriendly deltas like 0.1
+        let nextValue = parseFloat((parseFloat(value) + delta).toFixed(2));
+
+        // defaultProps won't work if the user passes in null, so just default
+        // to +/- infinity here instead, as a catch-all.
+        nextValue = Utils.clamp(nextValue, min || -Infinity, max || Infinity);
+
+        return nextValue.toString();
+    }
+
+    private getDelta(direction: number, e: React.MouseEvent<HTMLElement> | React.KeyboardEvent<HTMLElement>) {
+        const { majorStepSize, minorStepSize, stepSize } = this.props;
+
+        if (e.shiftKey && majorStepSize != null) {
+            return direction * majorStepSize;
+        } else if (e.altKey && minorStepSize != null) {
+            return direction * minorStepSize;
+        } else {
+            return direction * stepSize;
+        }
+    }
+
+    private isValueNumeric(value: string) {
+        // checking if a string is numeric in Typescript is a big pain, because
+        // we can't simply toss a string parameter to isFinite. below is the
+        // essential approach that jQuery uses, which involves subtracting a
+        // parsed numeric value from the string representation of the value. we
+        // need to cast the value to the `any` type to allow this operation
+        // between dissimilar types.
+        return value != null && ((value as any) - parseFloat(value) + 1) >= 0;
+    }
+
+    private getValueOrEmptyValue(props: INumericInputProps) {
+        return (props.value != null)
+            ? props.value.toString()
+            : NumericInput.VALUE_EMPTY;
+    }
+
+    private maybeInvokeOnUpdateCallback(value: string) {
+        if (this.props.onUpdate) {
+            this.props.onUpdate(value);
```

Comment:
Done (was seeing persistent type errors with Utils.safeInvoke at some point in the development of this component, hence these weird workarounds; those issues seem to have gone away)